### PR TITLE
RFC: Check for reproducible builds.

### PR DESCRIPTION
**Pull request by Gottox**
### Introduction

In void-packages the packages are anything but reproducible. Many other distributions, first and formost [NixOS](https://nixos.org/) and even [Debian](https://wiki.debian.org/ReproducibleBuilds) already did a lot of work to generate packages with stable checksums. Void's build system is able to do something similiar - with a few constraints - without much work.

### This is a starting point, not more.

As a first step to actually get an idea how bad the situation is I implemented a simple checker that compares the checksum of packages defined in templates to the actual result and spits out warnings they don't match.

This also introduces new variables to the templates:

`pkg_checksum_<arch>`, where <arch> is a sanitized version of the resulting architecture (`x86_64_musl` for `x86_64-musl` for example)

### Constraints:

* the packages are currently build with the githash backed in. This is an issue as the build is only stable within a certain commit.

### ToDo

* The documentation is currently not done, but will be added later.
* Find a way to make our package format reproducable across commits.

### CODE REVIEW

File: common/hooks/post-pkg/01-check-reproduce.sh [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+		arch=$XBPS_TARGET_MACHINE
+	else
+		arch=$XBPS_MACHINE
+	fi
+	binpkg=${pkgver}.${arch}.xbps
+
+	checksum_ptr="pkg_checksum_${arch//-/_}"
+	checksum_want=${!checksum_ptr}
+
+	checksum_have=$(sha256sum "$binpkg" | awk '{ print $1 }')
```
Comment (ericonr) [inline common/hooks/post-pkg/01-check-reproduce.sh:27]: It's easier to use `xbps-digest` :p

**Review by ericonr — COMMENTED** [review]
I'm against this, because our build system isn't deterministic.

- the `source-revisions` metadata depends on what commit the builder picked up to build the packages, and no one but people who can push to master can even guess at that value, but even then if the builder is stuck in some other build your guess won't be right.
- if someone pushes something while you push another thing, and that something affects your build, you will once again have issues.

Furthermore, how Debian and Arch do it is:

- archive all built packages (we could throw out the `-dbg` ones)
- dump build information for each package (we already have most of the build information because we have a monorepo and `source-revisions`, so a git checkout could be enough)
- have "rebuilder" servers which rebuild the universe/specific packages, using the exact environment the packages in the binary repository were built in, and comparing the results

The way I see it, the checksum for a package and its contents, when it comes to reproducibility, isn't a property of the source repository, but of the binary repository, and therefore it shouldn't be tracked here. The reproducible builds infrastructure should be separate.

**Comment by ericonr** [comment]
To explain my comment above, tracking `source-revisions` is equivalent to `.BUILDINFO` files used by other distros (which can be necessary because they don't use monorepos and so the build environment is much more variable), and so I'm strongly against its removal.

IIRC there was some discussion about whether a `.BUILDINFO` file should be included in the package itself or be a "side cart" file, but that adds complexity, more files to sign and track, and removes what is currently built-in information that can be useful for someone to try and reproduce things locally (without having to download the side cart with a weird path).

**Comment by ericonr** [comment]
Furthermore, this would only make sense for the builders at an exact moment in time. Someone bootstrapping from zero would have different checksums at a lot of stages.

**Comment by Chocimier** [comment]
1. I enjoy seeing effort to increase build reproducibility.
2. Ericonr is right that presence of `source-revisions` increases reproducibility, as it provides information on what dependencies, build tools and packaging tools, including xbps-src build-style and hooks, all affecting resulting package, was used. However guarantee of mismatching checksum of packages built from different commits, even if diff is unrelated to package, may still pose a problem.
3. Some packages will still come out identical even without reproducing every dependency recursively since foundation of distro. As I understand, this PR aims to assess is it closer 5 or 75 percent of packages.
4. How do you plan to use this hook? Is it filling variable locally, then observing hook messages on builder? Is it to store historical checksums in source repository? I am asking, because we already collect checksum of packages in binary repo, as sig files. If your workflow could be reversed to verify official signatures against packages reproduced outside of builders, then all packages are already checksummed, and this hook may be not necessary.

**Comment by ericonr** [comment]
> Some packages will still come out identical even without reproducing every dependency recursively since foundation of distro.

For the record, `libzstd` doesn't guarantee identical results when using different versions. The binary format is always compatible, but different choices can be made.

> I am asking, because we already collect checksum of packages in binary repo, as sig files. If your workflow could be reversed to verify official signatures against packages reproduced outside of builders, then all packages are already checksummed, and this hook may be not necessary.

I think having it external to void-packages would be better... Having it be a field in templates means bigger diffs for updates, manual work to build the package for all archs we want to check (plus the possibility of forgetting to update one field or another), and other things. If we could have some way of pushing our local checksums from testing updates to a separate repo, and then have daily CI for that repository that checks if hashes from locally built packages match what's listed in repodata, that'd be almost as useful...

Anyway, we still would need to find a way to deal with metadata... Maybe unpack the package files and run a content checksum, ignoring the metadata? It wouldn't help catch issues in metadata generation (which there are, sometimes the list of libraries goes in a different order iirc), but it would be something.

**Comment by ericonr** [comment]
So, thinking about this.

- the current version of this PR won't almost ever have matched checksums, because of `source-revisions`; we would have to work around that if we want to go with this path
- very few people build packages locally with debug symbols, and `compile with debug + strip` is different from `compile without debug`, so anyone updating templates will need to consider it
- void doesn't have a package archive, nor do we have `BUILDINFO`. While `source-revisions` provides the git repo state (hooks and build styles), and, therefore, the version of all the packages listed in the template in `(host)makedepends`, it doesn't record the version of the dependencies of the packages listed in `(host)makedepends`. Specifying a `BUILDINFO` file to record that information would be necessary. The archive would be necessary to have somewhere to fetch the specific versions of packages from

The current vision of `reproducible-builds` usually means trying to have the same build environment to see if the package is deterministic. If we want to study what packages are "reproducible" when built with different dependencies and etc (remembering that even line numbers in headers can change debug info), I think we would have to call this effort something else.

**Comment by Gottox** [comment]
> I'm against this, because our build system isn't deterministic.

We shouldn't improve deterministic builds because we don't have deterministic builds. I don't see how that's a valid reason.

**Comment by Gottox** [comment]
There are problems with changing metadata, but there are ways around this. We have all the moving parts available and can work on them.

**Comment by ericonr** [comment]
> We shouldn't improve deterministic builds because we don't have deterministic builds. I don't see how that's a valid reason.

The point is that the checksum won't match in 100% of the cases because of the metadata (specifically `source-revisions` - when you update the template with the new checksum for the package, that will change the commit hash and therefore the checksum; it's a loop). Most (all?) reproducible builds projects start from a binary artifact produced by the build and check if they can re-generate that based on information they had about the build environment, instead of checking if the builder produced the expected binary. I believe guix tries to check checksums from the bootstrap build process, but they always start from 0, which is a very different situation than what we have.

The determinism I was talking about wasn't in the sense of reproducible-builds, but in the sense that we can't control the value of `source-revisions` until the package has been built.

> There are problems with changing metadata, but there are ways around this. We have all the moving parts available and can work on them.

Which is why I'm proposing to study them. A content checksum (like our `@hash` checksum for templates) that skips `props.plist` would do most of the work (though it'd be slower). The issue then would be how to enforce policy to make people build packages for all archs, with the same settings used by builders (and then you have to take into account that `SOURCE_DATE_EPOCH` is derived from when a template was last touched, and that's also a loop of information, because a commit with a different date will once again alter the date, and this value *can* appear in normal files, for things like "build time"; this should be a rare case, though).

I don't think we have resources for a package archive or a "rebuilder" server that just tries to rebuild packages on its own, but I don't think the added workload and noise ("is this checksum wrong because someone forgot to udpate it or because something changed?") makes it worth having reproducible builds tracked in this repository. I would gladly contribute to a separate repository that tracks "verifications" from void users, and we could figure out some way to calculate "hashes" that skip the difference caused by building from different hashes (which would, unfortunately, partly defeat the purpose of reproducible-builds: to be able to verify that packages are exactly the same by simply comparing hashes from raw archives, without complicated tools).

To be clear, I am very interested in this work, and would like to see Void more reproducible (and have its reproducibility quantified as well). But I don't think this is a correct direction.

**Pull request closed without merging**